src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,109 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    static final int DEFAULT_CAPACITY = 10;
+    static final float INCREASE_CAPACITY_COEFFICIENT = 1.5f;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexCheck(index);
+        checkCapacity(1);
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] newElements = new Object[list.size()];
+        checkCapacity(list.size());
+        for (int i = 0; i < newElements.length; i++) {
+            newElements[i] = list.get(i);
+        }
+        System.arraycopy(newElements, 0, elementData, size, newElements.length);
+        size += newElements.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index + 1);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index + 1);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index + 1);
+        T element = (T) elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        size--;
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndexByValue(element);
+        return remove(index);
+    }
+
+    private int getIndexByValue(T value) {
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(value, elementData[i])) {
+                return i;
+            }
+        }
+        throw new NoSuchElementException("No such element");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void indexCheck(int index) {
+        if (index < 0 || index > elementData.length) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds");
+        }
+        if (index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+    }
```

Comment:
if we separate this methods < it will be better

Reply:
but in this way I can check two conditions using one method
and if I separate them I will call two methods instead of one every time

Reply:
I think we don't need `index > elementData.length` check
also, be careful, we have different constraints for `get(int index)` and `add(T value, int index)` methods